aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -756,9 +785,101 @@ Tensor dense_to_sparse_bsr(const Tensor& self, IntArrayRef blocksize) {
 }
 
 Tensor dense_to_sparse_bsc(const Tensor& self, IntArrayRef blocksize) {
-  AT_ERROR(
-      "Conversion from ", self.layout(), " to SparseBsc is currently not supported.");
-  return self;
+  TORCH_CHECK(
+      blocksize[0] > 0 && blocksize[1] > 0,
+      "blocksize needs to be non zero, but got ",
+      blocksize);
+  TORCH_CHECK(
+      self.size(-2) % blocksize[0] == 0,
+      "Tensor size(-2) ",
+      self.size(-2),
+      " needs to be divisible by blocksize[0] ",
+      blocksize[0]);
+  TORCH_CHECK(
+      self.size(-1) % blocksize[1] == 0,
+      "Tensor size(-1) ",
+      self.size(-1),
+      " needs to be divisible by blocksize[1] ",
+      blocksize[1]);
+  auto n_batch_dim = self.dim() - 2;
+  auto is_batched = n_batch_dim > 0;
+  auto values = _batch_tile_tensor(self, blocksize);
+  auto not_zero_mask = _batch_tile_tensor((self != 0), blocksize);
+  not_zero_mask = not_zero_mask.any(-1).any(-1);
+
+  if (is_batched) {
+    // Some extra setup and checks for batched
+    if (n_batch_dim > 1) {
+      // Flatten to single batch dim
+      values = values.flatten(0, n_batch_dim - 1);
+      not_zero_mask = not_zero_mask.flatten(0, n_batch_dim - 1);
+    }
+    TORCH_CHECK(
+        not_zero_mask.size(0) > 0,
+        "to_sparse_bsc: Expected product of batch dimensions to be non-zero.");
+
+    // Batched inputs must have the  same nnz per batch.
+    auto not_zero_mask_0 = not_zero_mask.select(0, 0);
+    auto nse_per_batch = not_zero_mask_0.sum().repeat(not_zero_mask.size(0));
+    TORCH_CHECK(
+        not_zero_mask.sum({-2, -1}).equal(nse_per_batch),
+        "Expect the same number of specified elements per batch.");
+
+    // We need to join batches into a matrix increasing the number of cols.
+    // This allows us to create indices for col compressed matrix and de-batch
+    // them later. Avoiding a kernel start for each batch.
+    values = values.transpose(0, 1).flatten(1, 2);
+    not_zero_mask = not_zero_mask.transpose(0, 1).flatten(1, 2);
+  }
+
+  Tensor col_indices;
+  Tensor row_indices;
+  // Compressed col indices are the same as the row indices of the transpose!
+  std::tie(row_indices, col_indices) = _not_zero_mask_to_col_row_indices(
+      not_zero_mask.transpose(1, 0), at::kLong, not_zero_mask.device());
+  // This only works if the col_indices vector is in ascending order.
+  Tensor ccol_indices = at::_convert_indices_from_coo_to_csr(
+      col_indices, not_zero_mask.size(-1), false /*out_int32*/);
+  {
+    // We need the values to blocked row major (they are), but blocks to be col
+    // major, so we transpose the leading two dims.
```

Comment:
So, it says that the blocks are row major but we want them to be col major, so that is why transpose?

Reply:
This is a bit confusing as there is the order of *blocks* and the order of *a block*.
The values array is a matrix with values that are matrices. Inner structure is what I am calling the blocked-values. 

While the blocks must be in col-major, and the mask must also, for the result's values array to be correct the individual blocked-value must remain in its current order. 


Suppose you have the following matrix 2x3 :
```   
T = [[0, 0, a],
     [0, b, 0]]
```
The computed `ccol_indices` would be `[0, 0, 1, 2]`
The and row indices `[1, 0]`. So we would expect the first element in `values` to be `b` and the second to be `a`. 

If we compute the final values here without a transpose:

The initial values are identical to T, and the mask would be 
```
M = [[F, F, T],
     [F, T, F]]
```

The flattened mask, converted to indices is `[2, 4]` and the final values array is `[a, b]`. If we transpose both before flattening we get the correct result back `[b, a]` 

For the blocked case, `a` and `b` are no longer scalars they are rectangular blocks with `blocksize` shape and `0` is a block of zeros with the same shape. The original tensor (T above) has `blocksize[0]*2` rows and `blocksize[1]*3` columns, but when we get to the point where we are indexing values to extract the nonzero values only, it has been tiled such that it looks the same as `T` above (with the new definitions of `0` `a` and `b`). 

`T` is row-major, and its blocks `a`, `b` are also row-major. We need to index a col-major T extracting the non-zero blocks in the correct order `[b, a]` but do not want to make each block itself col-major as well.